### PR TITLE
Emit table config from const entries

### PR DESCRIPTION
**Pull request by kamleshbhalui**
This PR emits new table config in txt format from const entries present in p4 program to be consumed by dpdk target.

i.e.

match 0x2/0x2 action MainControlImpl.tcp_syn_packet 
match 0x1/0x1 action MainControlImpl.tcp_fin_or_rst_packet 
match 0x4/0x4 action MainControlImpl.tcp_fin_or_rst_packet 

for p4 const entries:
const bit<8> TCP_RST_MASK = 0x04;
const bit<8> TCP_SYN_MASK = 0x02;
const bit<8> TCP_FIN_MASK = 0x01;

 const entries = {
            TCP_SYN_MASK &&& TCP_SYN_MASK: tcp_syn_packet;
            TCP_FIN_MASK &&& TCP_FIN_MASK: tcp_fin_or_rst_packet;
            TCP_RST_MASK &&& TCP_RST_MASK: tcp_fin_or_rst_packet;
        }

### CODE REVIEW

File: backends/dpdk/dpdkArch.cpp [hunk 1]
```diff
@@ -2448,4 +2448,111 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
     return m;
 }
 
+int EmitDpdkTableConfig::getTypeWidth(const IR::Type* type,
+                                      P4::TypeMap* typeMap) {
+    /* Treat error type as string */
+    if (type->is<IR::Type_Error>())
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.cpp:2454]: Iirc, Type_Error is treated as bit<8>

File: backends/dpdk/dpdkArch.h [hunk 2]
```diff
@@ -26,6 +27,8 @@ limitations under the License.
 #include "lib/ordered_map.h"
 #include "dpdkProgramStructure.h"
 #include "constants.h"
+#include "lib/gmputil.h"
+#include "frontends/p4/enumInstance.h"
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.h:31]: Why is this added?

File: backends/dpdk/dpdkArch.cpp [hunk 3]
```diff
@@ -2448,4 +2448,111 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
...
+            } else {
+                ::error(ErrorType::ERR_UNSUPPORTED,
+                        "%1% unsupported argument expression", arg);
+                continue;
+            }
+        }
+    }
+
+    for (size_t i = 0; i < argVals.size(); i++) {
+        dpdkTableConfigFile<<paramNames[i] << " "<< argVals[i]<<" ";
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.cpp:2528]: these print statements can be made as utility function, just taking a string as input. It will look cleaner

**Review by usha1830 — COMMENTED** [review]
Overall design looks fine to me

File: backends/dpdk/dpdkArch.cpp (hunk 1 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp:2454]: ok

File: backends/dpdk/dpdkArch.cpp (hunk 3 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp:2528]: ok

**Comment by kamleshbhalui** [comment]
table config for added testcase in this PR:

  match 0x1 0x1 0x2 action ingress.execute x 0x1
  match 0x1 0x1 0x2 action ingress.execute x 0x1
  match 0x1 0x3 0x3 action ingress.execute x 0x1

File: backends/dpdk/dpdkArch.cpp [hunk 4]
```diff
@@ -2448,4 +2448,242 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
...
+}
+
+big_int EmitDpdkTableConfig::convertSimpleKeyExpressionToBigInt(
+            const IR::Expression* k, int keyWidth, P4::TypeMap* typeMap) {
+    if (k->is<IR::Constant>()) {
+        return k->to<IR::Constant>()->value;
+    } else if (k->is<IR::BoolLiteral>()) {
+        return static_cast<big_int>(k->to<IR::BoolLiteral>()->value
+                                    ? 1 : 0);
+    } else if (k->is<IR::Member>()) {  // handle SerEnum members
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.cpp:2478]: This can be changed to if (auto mem = k->to<IR::Member>())

File: backends/dpdk/dpdkArch.cpp [hunk 5]
```diff
@@ -2448,4 +2448,242 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
...
+
+void EmitDpdkTableConfig::print(cstring str, cstring sep) {
+    dpdkTableConfigFile<<str<<sep;
+}
+
+void EmitDpdkTableConfig::print(big_int str, cstring sep) {
+    dpdkTableConfigFile<<"0x"<<std::hex<<str<<sep;
+}
+
+big_int EmitDpdkTableConfig::convertSimpleKeyExpressionToBigInt(
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.cpp:2471]: Does the added test cover all the if conditions? If not, can we add more entries covering all these cases

File: backends/dpdk/dpdkArch.cpp [hunk 6]
```diff
@@ -2448,4 +2448,242 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
...
+        std::vector<cstring> paramNames;
+        std::vector<big_int> argVals;
+        auto parameter = actionDecl->parameters->parameters.at(0);
+        auto type = typeMap->getType(parameter);
+        if (auto actArg = type->to<IR::Type_Struct>()) {
+            for (auto f : actArg->fields)
+                paramNames.push_back(f->name);
+        }
+        for (auto args : *actionCall->arguments) {
+            for (auto arg : args->expression->to<IR::ListExpression>()->components) {
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.cpp:2518]: Same here. Lets update the test to cover all these cases

File: backends/dpdk/dpdkArch.cpp [hunk 7]
```diff
@@ -2448,4 +2448,242 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
...
+            auto tableKey = table->getKey()->keyElements.at(keyIndex++);
+            auto keyWidth = getTypeWidth(tableKey->expression->type, typeMap);
+            auto matchType = getKeyMatchType(tableKey, refMap);
+            if (matchType == P4::P4CoreLibrary::instance.exactMatch.name) {
+              addExact(k, keyWidth, typeMap);
+            } else if (matchType == P4::P4CoreLibrary::instance.lpmMatch.name) {
+              addLpm(k, keyWidth, typeMap);
+            } else if (matchType == P4::P4CoreLibrary::instance.ternaryMatch.name) {
+              addTernary(k, keyWidth, typeMap);
+            } else if (matchType == P4V1::V1Model::instance.rangeMatchType.name) {
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.cpp:2662]: The range and optional match type should also be available in PSA. Do we not have such definitions for PSA? I think we should avoid using the v1model.h

File: backends/dpdk/dpdkArch.h [hunk 8]
```diff
@@ -987,6 +993,40 @@ class ElimHeaderCopy : public Transform {
     const IR::Node* postorder(IR::Member* m) override;
 };
 
+class EmitDpdkTableConfig : public Inspector {
```
Comment (usha1830) [inline backends/dpdk/dpdkArch.h:996]: We should  add a comment that this pass reuses the code from P4RuntimeEntriesConverter

**Review by usha1830 — CHANGES_REQUESTED** [review]

**Review by psivanup — APPROVED** [review]
Looks fine to me.

File: backends/dpdk/dpdkArch.h (hunk 8 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.h:997]: done

File: backends/dpdk/dpdkArch.cpp (hunk 7 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp]: done

**Review by usha1830 — APPROVED** [review]
Looks fine
